# Introduction

For this project, OilyGiant mining company is searching for the best region for a new well. Using the data on oil samples from three regions, I will build a model to help pick the region with the highest profit margin, then analyze potential profit and risks. 

The features are as follows: id — unique oil well identifier, f0, f1, f2 — three features of points (their specific meaning is unimportant, but the features themselves are significant).

The target is as follows: product — volume of reserves in the oil well (thousand barrels).

## Step 1

Importing all necessary libraries and modules. Preprocessing the data.

In [1]:
import pandas as pd
import numpy as np
from scipy import stats as st
import plotly.express as px 
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error



In [2]:
try:
    gd_0 = pd.read_csv('/System/Volumes/Data/Users/krystaltaylar/Desktop/TripleTen Projects/Sprint 9/geo_data_0.csv')
except FileNotFoundError:
    gd_0 = pd.read_csv('geo_data_0.csv')
    
try:
    gd_1 = pd.read_csv('/System/Volumes/Data/Users/krystaltaylar/Desktop/TripleTen Projects/Sprint 9/geo_data_1.csv')
except FileNotFoundError:
    gd_1 = pd.read_csv('geo_data_1.csv')

try:
    gd_2 = pd.read_csv('/System/Volumes/Data/Users/krystaltaylar/Desktop/TripleTen Projects/Sprint 9/geo_data_2.csv')
except FileNotFoundError:
    gd_2 = pd.read_csv('geo_data_2.csv')
    
# read datasets

In [3]:
display(gd_0.info())
display(gd_0.shape)
display(gd_0.sample(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

(100000, 5)

id        f0        f1        f2     product
18490  TBCZX  0.791891 -0.008148  4.126439  148.595641
74185  PtDTy  1.611976 -0.206749 -0.724736   58.892678
42333  EXBl5  0.196319  1.160043  5.585835   59.913868
69918  3617i  0.530577  0.774575  3.478925   97.526241
15989  A8N8q  1.794770 -0.291387  5.495769  159.910943
86443  eXzxR -0.249685  0.995128  4.026519   59.669947
13608  J1vpY -0.747116  0.641750 -0.080416   24.205354
48600  kVcWg -0.849000  0.032056  3.967369  130.864315
38884  tqLBC  1.221502 -0.388941  1.838478   72.993649
46020  eZI6Q  1.028638  0.231397  1.931040   77.963066

In [4]:
display(gd_1.info())
display(gd_1.shape)
display(gd_1.sample(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

(100000, 5)

id         f0         f1        f2     product
65737  GV63W  -8.277506   4.256231  1.999234   57.085625
3796   BISf3   9.332995  -0.141692  2.994690   80.859783
18592  XV4aV  14.833983  -1.026746  0.005034    0.000000
87182  qrFEj   3.805392  -0.120883  0.988387   26.953261
48899  Yi5EE  10.981053 -11.363496  1.006123   26.953261
46453  V2CiU   4.536915  -6.060011  4.989641  137.945408
79112  m7n7s  -6.545592  -9.942127  4.997426  137.945408
42005  htObm -12.518764 -11.613894 -0.003448    3.179103
81950  S2VGP  11.878952 -10.505874  3.992188  107.813044
26127  9nH31 -15.892690 -12.016477  4.999542  137.945408

In [5]:
display(gd_2.info())
display(gd_2.shape)
display(gd_2.sample(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

(100000, 5)

id        f0        f1        f2     product
45802  b8dNM  0.957119 -1.135107  7.478646  130.056189
26966  pTUfq -0.733959  1.276891  6.357212  153.717643
9824   God39 -2.295875 -2.477390  3.752415   33.563408
48986  CYalD  1.197518 -0.232797 -2.028930   38.754482
64576  37moJ -2.627670 -0.460912 -0.278532  101.266048
14437  zyo0J -1.003587 -1.761955  3.056210   78.155932
88713  9hLbn -1.438696 -0.229112 -0.368589   88.904997
55784  qSy9l  1.175126  0.727027  5.742154  143.832990
43403  GltG1  0.562616  2.248483  1.969330   55.175280
26829  amVTh  1.643551  0.134511  3.400927   30.259202

In [6]:
gd_0['id'].duplicated().sum()

np.int64(10)

In [7]:
duplicate_ids = gd_0[gd_0.duplicated(subset=['id'], keep=False)]
print(duplicate_ids.sort_values(by='id'))

          id        f0        f1         f2     product
66136  74z30  1.084962 -0.312358   6.990771  127.643327
64022  74z30  0.741456  0.459229   5.153109  140.771492
51970  A5aEY -0.180335  0.935548  -2.094773   33.020205
3389   A5aEY -0.039949  0.156872   0.209861   89.249364
69163  AGS9W -0.933795  0.116194  -3.655896   19.230453
42529  AGS9W  1.454747 -0.479651   0.683380  126.370504
931    HZww2  0.755284  0.368511   1.863211   30.681774
7530   HZww2  1.061194 -0.373969  10.430210  158.828695
63593  QcMuo  0.635635 -0.473422   0.862670   64.578675
1949   QcMuo  0.506563 -0.323775  -2.215583   75.496502
75715  Tdehs  0.112079  0.430296   3.218993   60.964018
21426  Tdehs  0.829407  0.298807  -0.049563   96.035308
92341  TtcGQ  0.110711  1.022689   0.911381  101.318008
60140  TtcGQ  0.569276 -0.104876   6.440215   85.350186
89582  bsk9y  0.398908 -0.400253  10.122376  163.433078
97785  bsk9y  0.378429  0.005837   0.160827  160.637302
41724  bxg6G -0.823752  0.546319   3.630479   93

In [8]:
gd_1['id'].duplicated().sum()

np.int64(4)

In [9]:
duplicate_ids = gd_1[gd_1.duplicated(subset=['id'], keep=False)]
print(duplicate_ids.sort_values(by='id'))

          id         f0         f1        f2     product
5849   5ltQ6  -3.435401 -12.296043  1.999796   57.085625
84461  5ltQ6  18.213839   2.191999  3.993869  107.813044
1305   LHZR0  11.170835  -1.945066  3.002872   80.859783
41906  LHZR0  -8.989672  -4.286607  2.009139   57.085625
2721   bfPNe  -9.494442  -5.463692  4.006042  110.992147
82178  bfPNe  -6.202799  -4.820045  2.995107   84.038886
47591  wt4Uk  -9.091098  -8.109279 -0.002314    3.179103
82873  wt4Uk  10.259972  -9.376355  4.994297  134.766305


In [10]:
gd_2['id'].duplicated().sum()

np.int64(4)

In [11]:
duplicate_ids = gd_2[gd_2.duplicated(subset=['id'], keep=False)]
print(duplicate_ids.sort_values(by='id'))

          id        f0        f1        f2     product
45404  KUPhW  0.231846 -1.698941  4.990775   11.716299
55967  KUPhW  1.211150  3.176408  5.543540  132.831802
11449  VF7Jo  2.122656 -0.858275  5.746001  181.716817
49564  VF7Jo -0.883115  0.560537  0.723601  136.233420
44378  Vcm5J -1.229484 -2.439204  1.222909  137.968290
95090  Vcm5J  2.587702  1.986875  2.482245   92.327572
28039  xCHr8  1.633027  0.368135 -2.378367    6.120525
43233  xCHr8 -0.847066  2.101796  5.597130  184.388641


In this step, I checked the shape, datatypes, and a sample of each dataset. I can see there are no missing entries after calling info on each dataset. I checked for duplicates and though there appears to be a number of duplicate IDs in each dataset, their corresponding f0, f1, f2, and product values are different. These could be distinct obbservations, therefore, I didn't want to do anything to remove them from the dataset.

## Step 2
Train and test the model for each region

Each dataset will be split into the training and validation sets at a 75:25 ratio. In this portion, I will calculate the average volume of predicted reserves and model RMSE.

In [12]:
# extract the features and the target
gd_0_features = gd_0.drop(['product', 'id'], axis=1)
gd_0_target = gd_0['product']

gd_0_features_train, gd_0_features_valid, gd_0_target_train, gd_0_target_valid = train_test_split(
    gd_0_features, gd_0_target, test_size=0.25, random_state=12345) # split 25% of data to make validation set

gd_0_model = LinearRegression()
gd_0_model.fit(gd_0_features_train, gd_0_target_train)
gd_0_predictions_valid = gd_0_model.predict(gd_0_features_valid)

gd_0_result = mean_squared_error(gd_0_target_valid, gd_0_predictions_valid) **0.5
print("RMSE of the Linear Regression Model on the gd_0 validation set:", gd_0_result)
avg_pred_gd_0 = gd_0_predictions_valid.mean()
print("Average Predicted Reserves:", avg_pred_gd_0)





RMSE of the Linear Regression Model on the gd_0 validation set: 37.5794217150813
Average Predicted Reserves: 92.59256778438035


In [13]:
# extract the features and the target
gd_1_features = gd_1.drop(['product', 'id'], axis=1)
gd_1_target = gd_1['product']

gd_1_features_train, gd_1_features_valid, gd_1_target_train, gd_1_target_valid = train_test_split(
    gd_1_features, gd_1_target, test_size=0.25, random_state=12345) # split 25% of data to make validation set

gd_1_model = LinearRegression()
gd_1_model.fit(gd_1_features_train, gd_1_target_train)
gd_1_predictions_valid = gd_1_model.predict(gd_1_features_valid)

gd_1_result = mean_squared_error(gd_1_target_valid, gd_1_predictions_valid) **0.5
print("RMSE of the Linear Regression Model on the gd_1 validation set:", gd_1_result)
avg_pred_gd_1 = gd_1_predictions_valid.mean()
print("Average Predicted Reserves:", avg_pred_gd_1)

RMSE of the Linear Regression Model on the gd_1 validation set: 0.893099286775617
Average Predicted Reserves: 68.728546895446


In [14]:
# extract the features and the target
gd_2_features = gd_2.drop(['product', 'id'], axis=1)
gd_2_target = gd_2['product']

gd_2_features_train, gd_2_features_valid, gd_2_target_train, gd_2_target_valid = train_test_split(
    gd_2_features, gd_2_target, test_size=0.25, random_state=12345) # split 25% of data to make validation set

gd_2_model = LinearRegression()
gd_2_model.fit(gd_2_features_train, gd_2_target_train)
gd_2_predictions_valid = gd_2_model.predict(gd_2_features_valid)

gd_2_result = mean_squared_error(gd_2_target_valid, gd_2_predictions_valid) **0.5
print("RMSE of the Linear Regression Model on the gd_2 validation set:", gd_2_result)
avg_pred_gd_2 = gd_2_predictions_valid.mean()
print("Average Predicted Reserves:", avg_pred_gd_2)

RMSE of the Linear Regression Model on the gd_2 validation set: 40.02970873393434
Average Predicted Reserves: 94.96504596800489


On average, Region 0 has the second highest amount of predicted reserves. It's RMSE reveals it's, on average, around 37 thousand barrels off from the actual reserves. Region 1 has the lowest amount of predicted reserves, but also the lowest RMSE. Region 2 has the highest amount of predicted reserves, but also the highest RMSE.

In [15]:
actual_mean_gd_0 = gd_0['product'].mean()
actual_mean_gd_1 = gd_1['product'].mean()
actual_mean_gd_2 = gd_2['product'].mean()

print(f"Actual Mean Reserves - gd_0: {actual_mean_gd_0}")
print(f"Actual Mean Reserves - gd_1: {actual_mean_gd_1}")
print(f"Actual Mean Reserves - gd_2: {actual_mean_gd_2}")

Actual Mean Reserves - gd_0: 92.50000000000001
Actual Mean Reserves - gd_1: 68.82500000000002
Actual Mean Reserves - gd_2: 95.00000000000004


#### Step 3
Prepare for profit calculation

In [16]:
# store all key values for calculation in separate values
budget = 100_000_000 # 100 million usd
wells = 200 # number of wells for development
barrel_rev = 4500 # 4500 per 1000 barrels, in usd

cost_per_well = budget/wells

print("Cost of developing one well (USD):", cost_per_well)

Cost of developing one well (USD): 500000.0


Calculate the volume of reserves sufficient for developing a new well without lossess. Compare the obtained value with the average volume of reserves in each region.

In [17]:
min_reserves_needed = cost_per_well/barrel_rev
print("Minimum reserves per well needed:", min_reserves_needed)

Minimum reserves per well needed: 111.11111111111111


In [18]:
# Compare break-even volume with average reserves in each region
regions_avg_reserves = [avg_pred_gd_0, avg_pred_gd_1, avg_pred_gd_2]

for i, avg_reserves in enumerate(regions_avg_reserves):
    print(f"Region {i}: Average predicted reserves = {avg_reserves:.2f} thousand barrels")
    if avg_reserves >= min_reserves_needed:
        print(f"Region {i} meets the break-even requirement.\n")
    else:
        print(f"Region {i} does NOT meet the break-even requirement.\n")

Region 0: Average predicted reserves = 92.59 thousand barrels
Region 0 does NOT meet the break-even requirement.

Region 1: Average predicted reserves = 68.73 thousand barrels
Region 1 does NOT meet the break-even requirement.

Region 2: Average predicted reserves = 94.97 thousand barrels
Region 2 does NOT meet the break-even requirement.



Each region is performing under the amount of reserves necessary for profit.

#### Step 4
Write a function to calculate profit from a set of selected oil wells and model predictions.

In [19]:
gd_0_predictions_valid = pd.Series(gd_0_predictions_valid).reset_index(drop=True)
gd_1_predictions_valid = pd.Series(gd_1_predictions_valid).reset_index(drop=True)
gd_2_predictions_valid = pd.Series(gd_2_predictions_valid).reset_index(drop=True)

gd_0_target_valid = gd_0_target_valid.reset_index(drop=True)
gd_1_target_valid = gd_1_target_valid.reset_index(drop=True)
gd_2_target_valid = gd_2_target_valid.reset_index(drop=True)

In [20]:
budget = 1000000
cost_point = 5000
points_budget = budget // cost_point
product_price = 45

def profit(target, predictions):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected_points = target[predictions_sorted.index][:points_budget]
    product = selected_points.sum()
    revenue = product * product_price
    cost = budget
    return revenue - cost

In [21]:
# Calculate profits for each region
region_profit_0 = profit(target=gd_0_target_valid, predictions=gd_0_predictions_valid)
region_profit_1 = profit(target=gd_1_target_valid, predictions=gd_1_predictions_valid)
region_profit_2 = profit(target=gd_2_target_valid, predictions=gd_2_predictions_valid)

print(f"Region 0 Profit:", region_profit_0)
print(f"Region_1 Profit:", region_profit_1)
print(f"Region_2 Profit:", region_profit_2)

# Store the region profits
region_profits = {
    'Region 0': region_profit_0,
    'Region 1': region_profit_1,
    'Region 2': region_profit_2
}

# Find the region with the highest profit
best_region = max(region_profits, key=region_profits.get)
best_profit = region_profits[best_region]

print(f"The best region for oil wells' development is {best_region} with a profit of {best_profit:.2f} USD.")

Region 0 Profit: 332082.604313985
Region_1 Profit: 241508.66966815107
Region_2 Profit: 271034.9963599832
The best region for oil wells' development is Region 0 with a profit of 332082.60 USD.


#### Step 5
Calculate risks and profit for each region:

In [22]:
target = gd_0_target_valid
predictions = gd_0_predictions_valid
region = 0
sample_size = 500
boostrap_size = 1000
state = np.random.RandomState(42)
profit_values = []
for i in range(boostrap_size):
    target_sample = target.sample(sample_size, replace=True, random_state=state)
    predictions_sample = predictions[target_sample.index]
    profit_values.append(profit(target_sample, predictions_sample))
profit_values = pd.Series(profit_values)
        
mean_profit = profit_values.mean()
confidence_interval = (profit_values.quantile(0.025), profit_values.quantile(0.975))
negative_profit_chance = (profit_values < 0).mean()

print("Mean profit =", mean_profit, "for region ", region)
print("95% confidence interval:", confidence_interval)
print("Risk of losses =", negative_profit_chance * 100, "%")

Mean profit = 40884.03060788564 for region  0
95% confidence interval: (np.float64(-12175.927995071554), np.float64(92761.9110715063))
Risk of losses = 6.6000000000000005 %


In [23]:
# Convert profit values to a DataFrame
df_profit_0 = pd.DataFrame({'Profit': profit_values})

fig = px.histogram(df_profit_0, x='Profit', title='Profit Distribution for Region 0', nbins=30)
fig.show()

In [24]:
target = gd_1_target_valid
predictions = gd_1_predictions_valid
region = 1
sample_size = 500
boostrap_size = 1000
state = np.random.RandomState(42)
profit_values = []
for i in range(boostrap_size):
    target_sample = target.sample(sample_size, replace=True, random_state=state)
    predictions_sample = predictions[target_sample.index]
    profit_values.append(profit(target_sample, predictions_sample))
profit_values = pd.Series(profit_values)
        
mean_profit = profit_values.mean()
confidence_interval = (profit_values.quantile(0.025), profit_values.quantile(0.975))
negative_profit_chance = (profit_values < 0).mean()

print("Mean profit =", mean_profit, "for region ", region)
print("95% confidence interval:", confidence_interval)
print("Risk of losses =", negative_profit_chance * 100, "%")

Mean profit = 50853.971237051956 for region  1
95% confidence interval: (np.float64(10018.773136972128), np.float64(94660.15058751087))
Risk of losses = 0.4 %


In [25]:
# Convert profit values to a DataFrame
df_profit_1 = pd.DataFrame({'Profit': profit_values})

fig = px.histogram(df_profit_1, x='Profit', title='Profit Distribution for Region 1', nbins=30)
fig.show()

In [ ]:
target = gd_2_target_valid
predictions = gd_2_predictions_valid
region = 2
sample_size = 500
boostrap_size = 1000
state = np.random.RandomState(42)
profit_values = []
for i in range(boostrap_size):
    target_sample = target.sample(sample_size, replace=True, random_state=state)
    predictions_sample = predictions[target_sample.index]
    profit_values.append(profit(target_sample, predictions_sample))
profit_values = pd.Series(profit_values)
        
mean_profit = profit_values.mean()
confidence_interval = (profit_values.quantile(0.025), profit_values.quantile(0.975))
negative_profit_chance = (profit_values < 0).mean()

print("Mean profit =", mean_profit, "for region ", region)
print("95% confidence interval:", confidence_interval)
print("Risk of losses =", negative_profit_chance * 100, "%")

Mean profit = 41822.96248648426 for region  2
95% confidence interval: (np.float64(-13790.339424570426), np.float64(95561.81276046397))
Risk of losses = 7.199999999999999 %


In [27]:
# Convert profit values to a DataFrame
df_profit_2 = pd.DataFrame({'Profit': profit_values})

fig = px.histogram(df_profit_2, x='Profit', title='Profit Distribution for Region 2', nbins=30)
fig.show()

# Conclusion

Based on the analysis of oil reserves and profit projections across the three regions, Region 0 emerges as the most profitable option, with an estimated profit of $332,082.60. However, none of the regions meet the break-even requirement of 111.11 thousand barrels per well, indicating potential financial risk in all cases.

While Region 1 has the lowest risk of losses (0.4%) and a relatively stable profit margin, its low average predicted reserves (68.73 thousand barrels) make it an unviable option. Region 2 has the highest average predicted reserves (94.97 thousand barrels) but comes with a higher risk of losses (7.2%).

Despite Region 0 being the most profitable, the risk of losses and the failure to meet the break-even point suggest that further investment in oil extraction should be carefully evaluated. A more in-depth economic feasibility study or alternative cost-cutting measures may be necessary before proceeding with oil well development.